Support for ML-DSA public key generation from private key

jakemas opened the pull request:
### Issues:
Resolves #CryptoAlg-2868

### Description of changes: 

It is often useful when serializing asymmetric key pairs to populate both the public and private elements, given only the private element. For this to be possible, an algorithm utility function is often provided to derive key material. ML-DSA does not support this in the reference implementation.

#### Background ML-DSA keypairs 

An ML-DSA private key is constructed of the following elements: (ref https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.204.pdf)
```
sk = (
      rho, // public random seed (32-bytes)
      tr,  // public key hash (64-bytes)
      key, // private random seed (32-bytes) (utilized during sign)
      t0,  // polynomial vector: encodes the least significant bits of public-key polynomial t, facilitating certain computational efficiencies.
      s1,  // secret polynomial vectors. These vectors contain polynomials with coefficients in a specified range, 
      s2.  // serving as the secret components in the lattice-based structure of ML-DSA.
)
```

An ML-DSA public key is constructed of the following elements:
```
pk = (
      rho, // public random seed (32-bytes)
      t1.  // compressed representation of the public key polynomial 
)
```

- The vector t is decomposed into two parts:
- `t1`: Represents the higher-order bits of `t`.
- `t0`: Represents the lower-order bits of `t`.

One can see that to reconstruct the public key from the private key, one must:
1. Extract all elements from `sk`, using the existing function in `/ml_dsa_ref/packing.c`: `ml_dsa_unpack_sk`
    1. This will provide `sk = (rho, tr, key, t0, s1, s2)`.
2. Reconstruct `A` using `rho` with the existing function in `/ml_dsa_ref/polyvec.c`: `ml_dsa_polyvec_matrix_expand`
3. Reconstruct `t` from `t = A*s1 + s2`
4. Drop `d` lower bits from `t` to get `t1`
5. Pack `rho`, `t1` into public key.
6. Verify `pk` matches expected value, by comparing SHAKE256(pk) + `tr` (unpacked from secret key).

This has been implemented in `ml_dsa_pack_pk_from_sk` -- not tied to the name, just using what I've seen so far in common nomenclature.

As the values of `d` differ for each parameter set of ML-DSA, we must create packing functions for each parameter size. As such, `ml_dsa_44_pack_pk_from_sk``, `ml_dsa_65_pack_pk_from_sk``, and `ml_dsa_87_pack_pk_from_sk`` have been added to `ml_dsa.h` to serve as utility functions in higher level EVP APIs.

### Call-outs:

The scope of this PR is only the algorithm level, using these functions for useful tasks such as populating the public key automatically on private key import -- will be added in subsequent PRs.

### Testing:
A new test has been added to `PQDSAParameterTest`, namely, `KeyConsistencyTest` that will assert that packing the key is successful, and that the key produced matches the original public key.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

justsmth reviewed (approved):
Thanks for getting this change up so quickly!

nebeid reviewed (commented):
May I suggest renaming to `pack_pk_from_sk`? I understand they're internal functions but they're exported and so renaming them later may be more difficult.

Nit: comments in https://github.com/aws/aws-lc/blob/4fbcd100235278e81320704c4f65c42eeaa5410f/crypto/ml_dsa/ml_dsa_ref/packing.c#L182-L187 should not have `const`.

jakemas commented:
> May I suggest renaming to `pack_pk_from_sk`? I understand they're internal functions but they're exported and so renaming them later may be more difficult.

Thank you, changed to `ml_dsa_pack_pk_from_sk` in [bf01ead](https://github.com/aws/aws-lc/pull/2142/commits/bf01ead882989d5a5b27c72d15cfbbe16a85a172).

> 
> Nit: comments in
> 
> https://github.com/aws/aws-lc/blob/4fbcd100235278e81320704c4f65c42eeaa5410f/crypto/ml_dsa/ml_dsa_ref/packing.c#L182-L187
> 
> should not have `const`.

Got them, also fixed in
[bf01ead](https://github.com/aws/aws-lc/pull/2142/commits/bf01ead882989d5a5b27c72d15cfbbe16a85a172)

vdukhovni commented on crypto/ml_dsa/ml_dsa_ref/packing.c:
```diff
@@ -2,6 +2,63 @@
+  ml_dsa_pack_pk(params, pk, rho, &t1);
+
+  // we hash pk to reproduce tr, check it with unpacked value to verify
+  SHAKE256(pk, params->public_key_bytes, tr_validate, ML_DSA_TRBYTES);
```
Another sanity check could be to verify that the just recomputed $`t_0`$ matches the original $`t_0`$ from the private key. If it doesn't, the original private key is invalid or the computation malfunctioned.
